### ISSUE TITLE
Fix S1944 FP: Rule fails to check for generic types

### ISSUE THREAD

**Issue by mtsfoni** [SonarSource/sonar-dotnet#9374]
### Description

Rule S1944: "Invalid casts should be avoided" wrongly claims that there are no types that extend the source and the target of the cast, when working with generic types.

### Repro steps

Having those types:

New Console App:
```
Result<string> error = new Error<string>();
var castedError = (IFailure)error; // <- Review this cast; in this project there's no type that extends 'Result<string>' and implements 'IFailure'.	

public abstract class Result<TData> { }

public interface IFailure { }

public class Error<TData> : Result<TData>, IFailure { }
```

Case was simplified for reproduction.

### Expected behavior

No warning, the cast is fine and doesn't throw an exception.

### Actual behavior

Warning.

### Known workarounds

Suppress the warning.

### Related information

The code that searches for a type that has source and destination type in common needs to consider generics better. 
After all, it only asks to review, so that's OK.

* Dotnet 8
* Visual Studio 2022 Community
* SonarAnalyzer.CSharp Version=9.26.0.92422

**Comment by zsolt-kolbay-sonarsource** [comment]
Thank you for reporting this issue. It's indeed a False Positive. 
It won't be easy to fix, as it requires Symbolic Execution (and the rule is listed under Symbol Execution rules but doesn't use the CFG).

**Comment by pavel-mikula-sonarsource** [comment]
It's not a SE rule in the new engine.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1580](https://sonarsource.atlassian.net/browse/NET-1580)

[NET-1580]: https://sonarsource.atlassian.net/browse/NET-1580?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
